cpp/include/cuspatial/experimental/spatial_window.cuh

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/vec_2d.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+
+#include <thrust/iterator/iterator_traits.h>
+
+namespace cuspatial {
+
+/**
+ * @brief Count of points (x,y) that fall within a rectangular query window.
+ *
+ * @ingroup spatial_relationship
+ *
+ * A point (x, y) is in the window if `x > window_min_x && x < window_max_x && y > window_min_y &&
+ * y < window_max_y`.
+ *
+ * Swaps `window_min.x` and `window_max.x` if `window_min.x > window_max.x`.
+ * Swaps `window_min.y` and `window_max.y` if `window_min.y > window_max.y`.
+ *
+ * @param[in] window_min lower-left (x, y) coordinate of the query window
+ * @param[in] window_max upper-right (x, y) coordinate of the query window
+ * @param[in] points_first beginning of range of (x, y) coordinates of points to be queried
+ * @param[in] points_last end of range of (x, y) coordinates of points to be queried
+ * @param[in] stream: The CUDA stream on which to perform computations
+ *
+ * @tparam InputIt Iterator to input points. Must meet the requirements of
+ * [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible.
+ * @tparam T The underlying coordinate type. Must be a floating-point type.
+ *
+ * @pre All iterators must have the same `value_type`, with the same underlying floating-point
+ * coordinate type (e.g. `cuspatial::vec_2d<float>`).
+ *
+ * @returns The number of input points that fall within the specified window.
+ */
+template <class InputIt, class T>
+typename thrust::iterator_traits<InputIt>::difference_type count_points_in_spatial_window(
+  vec_2d<T> window_min,
+  vec_2d<T> window_max,
+  InputIt points_first,
+  InputIt points_last,
+  rmm::cuda_stream_view stream = rmm::cuda_stream_default);
+
+/**
+ * @brief Find all points (x,y) that fall within a rectangular query window.
```

Comment:
Specify cartesian coordinates, I think.

Reply:
Down the road this, along with pip, would be really effective if they took a coordinate system pair as arguments.

Reply:
Actually I don't think we have to limit this to cartesian coordinates. This is effectively a 2D range query. There's no reason you can't use lonlat points and a 2D lonlat range.

Reply:
(In the future we may want to replace this API with a generic ND range query API...)

Reply:
I don't think that's strictly true - lon/lat is not cartesian. There exist points along a line of longitude that fall inside of the spherical "rectangle" defined by a lower left and upper right pair of coordinates that do not fall inside of the cartesian rectangle that is defined by those same coordinates.

Reply:
The API doesn't specify that the points are in any projection -- they can be Cartesian, Lon/Lat, or anything else, as long as the query window and the points are coordinates in the same dimensions.  This is a good reason to replace this with an API called "points_in_range" or something like that.

Reply:
I've renamed the APIs and clarified the docs. See commit #[88e7ee1](https://github.com/rapidsai/cuspatial/pull/579/commits/88e7ee18c9db8d9662a237f07e3e0a55bc1d8198) and after.